contracts/scripts/UpgradeShell.sol

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+pragma solidity 0.8.23;
+
+import {WETH9} from "canonical-weth/WETH9.sol";
+import {Script} from "forge-std/Script.sol";
+import {BeefyClient} from "../src/BeefyClient.sol";
+
+import {IGateway} from "../src/interfaces/IGateway.sol";
+import {IShell} from "../src/interfaces/IShell.sol";
+import {GatewayProxy} from "../src/GatewayProxy.sol";
+import {Gateway} from "../src/Gateway.sol";
+import {MockGatewayV2} from "../test/mocks/MockGatewayV2.sol";
+import {Agent} from "../src/Agent.sol";
+import {AgentExecutor} from "../src/AgentExecutor.sol";
+import {ChannelID, ParaID, OperatingMode} from "../src/Types.sol";
+import {SafeNativeTransfer} from "../src/utils/SafeTransfer.sol";
+import {stdJson} from "forge-std/StdJson.sol";
+import {UD60x18, ud60x18} from "prb/math/src/UD60x18.sol";
+
+function mDot(uint32 value) pure returns (uint128) {
+    // 1 mDOT = 0.001 DOT
+    return value * (10 ** 7);
+}
+
+function dot(uint32 value) pure returns (uint128) {
+    return value * (10 ** 10);
+}
+
+contract UpgradeShell is Script {
+    using SafeNativeTransfer for address payable;
+    using stdJson for string;
+
+    struct Config {
+        address gatewayProxy;
+        address beefyClient;
+        ParaID bridgeHubParaID;
+        bytes32 bridgeHubAgentID;
+        uint8 foreignTokenDecimals;
+        uint128 maxDestinationFee;
+        Gateway.Config initializerParams;
+    }
+
+    function readConfig() internal pure returns (Config memory config) {
+        config = Config({
+            gatewayProxy: 0x27ca963C279c93801941e1eB8799c23f407d68e7,
+            beefyClient: address(0),
+            bridgeHubParaID: ParaID.wrap(1002),
+            bridgeHubAgentID: 0x03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314,
+            foreignTokenDecimals: 10,
+            maxDestinationFee: dot(2),
+            initializerParams: Gateway.Config({
+                mode: OperatingMode.Normal,
+                deliveryCost: mDot(100), // 0.1 DOT
+                registerTokenFee: 0.002 ether,
+                assetHubParaID: ParaID.wrap(1000),
+                assetHubAgentID: 0x81c5ab2571199e3188135178f3c2c8e2d268be1313d029b30f534fa579b69b79,
+                assetHubCreateAssetFee: mDot(100), // 0.1 DOT
```

Comment:
Hey I calculated this to `20.19` DOT.

1. https://github.com/polkadot-fellows/runtimes/blob/main/system-parachains/asset-hubs/asset-hub-polkadot/src/lib.rs#L362
2. https://github.com/polkadot-fellows/runtimes/blob/main/system-parachains/asset-hubs/asset-hub-polkadot/src/lib.rs#L288
3. https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/system-parachains/constants/src/polkadot.rs#L68
4. https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/relay/polkadot/constants/src/lib.rs#L36

Python implementation:
```py
UNITS=10_000_000_000
DOLLARS=UNITS
CENTS=DOLLARS/100
MILLICENTS=CENTS/1_000

def deposit(items, bytes):
    return items * 20 * DOLLARS + bytes * 100 * MILLICENTS

deposit(1, 190) / 10**10 # 10 decimal places
```

I remember you saying that they are dropping the fee but I cannot find a related issue or PR.

Reply:
According to my understanding, this is really just bad naming in the Gateway contract which we should fix.

It turns out that `registerTokenFee` actually covers the token registration fee, and is credited to the AssetHub agent/sovereign after exchange rate conversions.

The `assetHubCreateAssetFee` item just covers the XCM execution fee on AssetHub

For example, in BridgeHub, the registration fee is added to `assetHubCreateAssetFee` on line https://github.com/paritytech/polkadot-sdk/blob/d237adfb13ceab8724fabee74469788e19e0e6af/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L167,

Reply:
> https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/relay/polkadot/constants/src/lib.rs#L36

That's relay chain config, AssetHub has different storage deposit config.

Reply:
Can see on Polkadot-JS that the asset registration deposit 0.2 DOT:
<img width="1137" alt="Screenshot 2024-05-16 at 02 29 01" src="https://github.com/Snowfork/snowbridge/assets/117534/e2ec78a1-685c-42c1-8cc6-9e54fcf561ec">

Reply:
Ok but then `registerTokenFee` is too low. `0.002 ETH` (6 USD).

`CreateForeignAssetDeposit` is set to `deposit(1, 190)` which is `20.19` DOT.

https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/system-parachains/asset-hubs/asset-hub-polkadot/primitives/src/lib.rs#L55

Asset hub just defers to relaychain config.
https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/system-parachains/constants/src/polkadot.rs#L68

And this test ensures that its the set the same as asset hub.
https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/system-parachains/asset-hubs/asset-hub-polkadot/src/lib.rs#L1748

Reply:
Ah ok. There is an extra divide by 100 that I missed here:
https://github.com/polkadot-fellows/runtimes/blob/d63ed882383e5c972757182e1125818666c9d660/system-parachains/constants/src/polkadot.rs#L68

Reply:
Yeah, so it's about 0.2 DOT(0.000464 ETH), then the `registerTokenFee` as 0.002 should be pretty enough.